analyzers/src/SonarAnalyzer.VisualBasic/Rules/UnusedStringBuilder.cs

```diff
@@ -0,0 +1,53 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Text;
+
+namespace SonarAnalyzer.Rules.VisualBasic;
+
+[DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+public sealed class UnusedStringBuilder : UnusedStringBuilderBase<SyntaxKind, VariableDeclaratorSyntax, InvocationExpressionSyntax, ReturnStatementSyntax>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => VisualBasicFacade.Instance;
+
+    protected override string GetVariableName(VariableDeclaratorSyntax declaration) => declaration.Names.FirstOrDefault().ToString();
+
+    protected override bool NeedsToTrack(VariableDeclaratorSyntax expression, SemanticModel semanticModel) =>
+        expression.Initializer is not null
+        && expression.Initializer.Value is ObjectCreationExpressionSyntax objectCreation
+        && objectCreation.Type.ToString().Equals(nameof(StringBuilder));
+
+    protected override SyntaxNode GetAncestorBlock(VariableDeclaratorSyntax declaration) =>
+        declaration.Ancestors().OfType<MethodBlockBaseSyntax>().First();
+
+    protected override bool IsIsStringInvoked(string variableName, IList<InvocationExpressionSyntax> invocations) =>
+        invocations.Any(x => x.Expression is MemberAccessExpressionSyntax { } member
```

Comment:
This propably doesn't work for `sb?.ToString();`

Reply:
This Null-conditional operator was a bit of a nightmare to support. We cannot just move downwards to the descendant nodes of an expression to find the local reference but we need to move toward the ancestors. Anyway, it is now supported 👍

Reply:
I know. It is bad, like really bad. I think it has to do with the left and right associativity of the `.` vs the `?.` operators but don't ask me about details. See also https://github.com/SonarSource/sonar-dotnet/pull/6230